### PR TITLE
Improve case insensitivity consistency

### PR DESCRIPTION
**Pull request by CAD97**
# Description

Add an extension trait `IgnoreCaseExt` to nu_utils which adds some case insensitivity helpers, and use them throughout nu to improve the handling of case insensitivity. Proper case folding is done via unicase, which is already a dependency via mime_guess from nu-command.

In actuality a lot of code still does `to_lowercase`, because unicase only provides immediate comparison and doesn't expose a `to_folded_case` yet. And since we do a lot of `contains`/`starts_with`/`ends_with`, it's not sufficient to just have `eq_ignore_case`. But if we get access in the future, this makes us ready to use it with a change in one place.

Plus, it's clearer what the purpose is at the call site to call `to_folded_case` instead of `to_lowercase` if it's exclusively for the purpose of case insensitive comparison, even if it just does `to_lowercase` still.

# User-Facing Changes

- Some commands that were supposed to be case insensitive remained only insensitive to ASCII case (a-z), and now are case insensitive w.r.t. non-ASCII characters as well.

# Tests + Formatting

- :green_circle: `toolkit fmt`
- :green_circle: `toolkit clippy`
- :green_circle: `toolkit test`
- :green_circle: `toolkit test stdlib`

### CODE REVIEW

File: crates/nu-command/src/conversions/into/bool.rs [hunk 1]
```diff
@@ -116,13 +116,13 @@ fn into_bool(
...
 fn string_to_boolean(s: &str, span: Span) -> Result<bool, ShellError> {
-    match s.trim().to_lowercase().as_str() {
+    match s.trim().to_ascii_lowercase().as_str() {
         "true" => Ok(true),
         "false" => Ok(false),
         o => {
             let val = o.parse::<f64>();
             match val {
-                Ok(f) => Ok(f.abs() >= f64::EPSILON),
+                Ok(f) => Ok(f != 0.0),
```
Comment (CAD97) [inline crates/nu-command/src/conversions/into/bool.rs:125]: Comparing to `f64::EPSILON` is almost never what you actually want to do. Yes, when comparing floats, you want to do so with *an* epsilon because of floating point imprecision, but the machine epsilon is an astoundingly bad choice for a comparison epsilon. There is no "perfect" comparison epsilon, and the comparison epsilon really wants to be relative to the scale of the compared values, due to the floating nature of floating point arithmetic.

A comparison to machine epsilon is in most cases no different to a direct comparison with `0.0`, except for a false appearance of satisfying a rule of "no float `==`" just for the purpose of satisfying the rule. Equality with `0` is both the value it makes most sense to directly compare to ("did they write a literal zero") and more honest than comparing with machine epsilon.

Yes, the clippy lint suggested comparing with `f64::EPSILON` for a long time. This was unfortunately bad advice that took an unfortunately long period to replace, due to there not being any straightforward advice to replace it with. You essentially just can't check floats for "effective equality" without domain specific knowledge.

Comment (ChrisDenton) [reply crates/nu-command/src/conversions/into/bool.rs:125]: Some recent rustlang discussion: https://github.com/rust-lang/rust/issues/116916

Comment (CAD97) [reply crates/nu-command/src/conversions/into/bool.rs:125]: And on clippy: https://github.com/rust-lang/rust-clippy/issues/6816

**Comment by fdncred** [comment]
Some of the feedback I'm getting from the nushell core-team is that lowercasing only ascii characters may be a mistake. Thoughts?

**Comment by kubouch** [comment]
It seems strange to me that some `to_lowercase()` calls are replaced with `to_ascii_lowercase()`. Shouldn't these be kept as `to_lowercase()`, or replaced with the new `to_folded_case()`?

**Comment by sholderbach** [comment]
For comparisons against known ASCII literals like in the config that is fine and gives a known capacity and is cheaper.

**Comment by CAD97** [comment]
That's accurate. Every case I switched to `to_ascii_lowercase` should have one side of the comparison known to be an ascii-only constant. If it isn't, that's an accidentally introduced bug. Someone scanning through to validate this assertion would be a useful review.

It's useful to use `to_ascii_lowercase` instead of `to_lowercase` or `to_folded_case` for multiple reasons:

- Performance: ASCII lowercase is better than lowercase is better than case folding.
- Intent: When comparing against a case constant, only normalizing for ASCII case more directly corresponds to the actual comparison being done. (I would have used `eq_ignore_ascii_case` more if not for usually `match`ing or passing to another function.)
- Correctness: While case folding *is* an aggressive lower casing, it shouldn't be necessary to know that to see that the code is doing the right thing. (Also, should `maß` compare equal to `mass`? It might, given case folding).

I can add comments to each instance of `to_ascii_Xcase` categorizing why that's sufficient if that's considered beneficial. Ask and it'll get done (sooner; I'll likely eventually end up doing it anyway to help make the PR easier to accept if it doesn't get merged first).

Side note: unicase merged the PR for exposing its `to_folded_case`, so when/if that gets published we can change ours to do a case fold instead of just lower case.

Side note: I'm considering the value of a follow-up PR using icu_collator for proper collation (sort) order (e.g. `á` sorting next to `a`), but that's a much bigger question to ask since 1. it's a new dependency, unlike std or unicase, 2. it requires deciding [how to handle ICU4X data](https://docs.rs/icu_provider/latest/icu_provider/constructors/index.html)[^icu4x], and 3. collation is locale sensitive (but you can use `und` for the most locale neutral option) (but it does natively provide numeric/natural sort order).

[^data]: The "best" approach is probably some combination of minimally sufficient always available compiled in data (the `und` locale) and then configuring paths to check for locale (or updated) data before using the fallback. Interesting would be communicating when the configured data is used and when built-in data is used, since there's essentially no way we get the *entire Rust strings software stack* to be generic over an ICU4X provider, so there'll always be *something* in the tree that uses baked Unicode data; it's just massively more convenient, to the point that ICU4X has it on by default.

**Comment by fdncred** [comment]
Thanks for the explanation @CAD97. We appreciate your help and support.

> I can add comments to each instance of to_ascii_Xcase categorizing why that's sufficient if that's considered beneficial.

Thanks for offering. I think this would be a good idea and helpful because I, for one, wouldn't have understood what you explained above.

**Comment by kubouch** [comment]
@CAD97 Thanks for the explanation, it makes sense to use ASCII lowercasing in cases where one side is known ASCII. Adding comments before each `to_ascii_lowercase()` would be helpful for the case when someone like me comes along and tries to “fix” it :laughing: . Alternatively,i nstead of adding comments, you could consider adding a `to_kown_ascii_lowercase()` (which would just call `to_ascii_lowercase()`) method to your trait and add a doc comment there explaining when it is supposed to be used. The latter would be more future-proof.

The Unicode string comparisons are tricky, so I think moving it into one place (the trait you introduced) is a good move IMO. Remaining questions, like “maß” vs. “mass”, would also be easier addressed within the trait.

File: crates/nu-command/src/conversions/into/bool.rs (hunk 1 above)
Comment (kubouch) [reply crates/nu-command/src/conversions/into/bool.rs:125]: Would you mind moving the float comparison into a separate PR? It is an issue on its own, it would be better addressed separately, IMO.

File: crates/nu-utils/src/casing.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+impl IgnoreCaseExt for str {
+    fn to_folded_case(&self) -> String {
+        // we only do to_lowercase, as unicase doesn't expose its case fold yet
+        // (seanmonstar/unicase#61) and we don't want to pull in another table
+        self.to_lowercase()
+    }
+
+    fn eq_ignore_case(&self, other: &str) -> bool {
+        UniCase::new(self) == UniCase::new(other)
+    }
```
Comment (sholderbach) [inline crates/nu-utils/src/casing.rs:50]: Would it make sense to have a newtype for which we implement `IgnoreCaseExt` so we can convert/fold our needles once before going through the haystacks? Else we can always go through the `to_folded_case` and the standard comparisons but that allocates for both the needle and each straw in the haystack.

File: crates/nu-cli/src/completions/variable_completions.rs [hunk 3]
```diff
@@ -44,9 +44,7 @@ impl Completer for VariableCompletion {
     ) -> Vec<Suggestion> {
         let mut output = vec![];
         let builtins = ["$nu", "$in", "$env"];
-        let var_str = std::str::from_utf8(&self.var_context.0)
-            .unwrap_or("")
-            .to_lowercase();
+        let var_str = std::str::from_utf8(&self.var_context.0).unwrap_or("");
```
Comment (sholderbach) [inline crates/nu-cli/src/completions/variable_completions.rs:47]: Am I reading this correctly that the variable completions will become case sensitive?

File: crates/nu-command/src/filters/sort.rs [hunk 4]
```diff
@@ -220,22 +221,22 @@ fn sort_record(
...
+            right_res.to_folded_case()
         } else {
             right_res
         };
 
         if natural {
             compare_str(left, right)
         } else {
-            left.partial_cmp(&right).unwrap_or(Ordering::Equal)
+            left.cmp(&right)
```
Comment (sholderbach) [inline crates/nu-command/src/filters/sort.rs:239]: Going for the correct `.cmp` shouldn't introduce any changes as `str: Ord + PartialOrd` the whole time, right?

File: crates/nu-utils/src/casing.rs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    /// additional changes to the source text to help make case folding
+    /// language-invariant and consistent. Case folded text should be used
+    /// solely for processing and generally should not be stored or displayed.
+    ///
+    /// Note: this method might only do [`str::to_lowercase`] instead of a
+    /// full case fold, depending on how Nu is compiled. You should still
+    /// prefer using this method for generating case-insensitive strings,
+    /// though, as it expresses intent much better than `to_lowercase`.
+    ///
+    /// [case folded]: <https://unicode.org/faq/casemap_charprop.html#2>
```
Comment (sholderbach) [inline crates/nu-utils/src/casing.rs:17]: Love the professional doccomments!

File: crates/nu-command/src/filters/sort.rs [hunk 6]
```diff
@@ -220,22 +221,22 @@ fn sort_record(
             b.0.clone()
         };
 
-        // Convert to lowercase if case-insensitive
+        // Fold case if case-insensitive
         let left = if insensitive {
-            left_res.to_ascii_lowercase()
+            left_res.to_folded_case()
```
Comment (sholderbach) [inline crates/nu-command/src/filters/sort.rs:226]: So this will be a breaking change for the `sort` command

File: crates/nu-command/src/sort_utils.rs [hunk 7]
```diff
@@ -125,28 +126,24 @@ pub fn sort(
                 if insensitive {
                     let span_a = a.span();
                     let span_b = b.span();
-                    let lowercase_left = match a {
-                        Value::String { val, .. } => {
-                            Value::string(val.to_ascii_lowercase(), span_a)
```
Comment (sholderbach) [inline crates/nu-command/src/sort_utils.rs:130]: Also anything depending on `crates/nu-command/src/sort_utils.rs` may incur a breaking change.

**Review by sholderbach — COMMENTED** [review]
Gave it a top down reading, the places where you narrowed it to ascii casing look good to me.

I tried to note what breaking changes to the behavior may result, so far I can only note `sort` and `sort-by` that may sort differently when operating in case-insensitive mode.

File: crates/nu-cli/src/completions/variable_completions.rs (hunk 3 above)
Comment (CAD97) [reply crates/nu-cli/src/completions/variable_completions.rs:47]: I believe `var_str` is only used to compare against `$nu` and `$env`. That comparison is now case sensitive, but when I tested I wasn't getting completions on `$Env.<tab>` anyway. This one could go either way tbh.

Comment (sholderbach) [reply crates/nu-cli/src/completions/variable_completions.rs:47]: Ah this is the logic for the subpaths of those. Can confirm your observation and it should still handle `$E<tab>` to `$env` etc. 

I am fine with that.

File: crates/nu-utils/src/casing.rs (hunk 2 above)
Comment (CAD97) [reply crates/nu-utils/src/casing.rs:50]: The main reason I didn't do so is that most of those cases are potentially using `contains` or `starts_with` or otherwise using operations which expect to have `&str` access (such as for natural sort). Plus, as discussed on the trait docs, for repeated comparisons it can often be preferable to fold once up front, since the folding process isn't exactly free either. ICU4X doesn't even offer case-folded equality comparison. (It does offer one-off collation, but that might still allocate internally to create sort keys.)

In short, the extra ceremony required to thread through a newtype isn't really worth it. A `Cow<str>` version which doesn't allocate if the string is already case folded might be beneficial, but no case folding provider exposes that possibility currently.

File: crates/nu-command/src/filters/sort.rs (hunk 4 above)
Comment (CAD97) [reply crates/nu-command/src/filters/sort.rs:239]: Correct. `.partial_cmp` won't (shouldn't) ever return `None` for (correctly implemented) `Ord` types.

File: crates/nu-glob/src/lib.rs [hunk 8]
```diff
@@ -1016,29 +1016,19 @@ fn in_char_specifiers(specifiers: &[CharSpecifier], c: char, options: MatchOptio
...
-                        .expect("internal error: getting start uppercase");
-                    let end_up = end
-                        .to_uppercase()
-                        .next()
-                        .expect("internal error: getting end uppercase");
-
                     // only allow case insensitive matching when
                     // both start and end are within a-z or A-Z
-                    if start != start_up && end != end_up {
+                    if start.is_ascii_alphabetic() && end.is_ascii_lowercase() {
```
Comment (CAD97) [inline crates/nu-glob/src/lib.rs:1021]: Uh whoops 🙃


```suggestion
                    if start.is_ascii_alphabetic() && end.is_ascii_alphabetic() {
```

File: crates/nu-utils/src/casing.rs (hunk 2 above)
Comment (sholderbach) [reply crates/nu-utils/src/casing.rs:50]: OK makes sense, thanks for the detailed explanation.

Let's go with the definite correctness improvements this all already provides. (If profiling tells us anything egregious we can still ponder particular implementations in the future anyways)

File: crates/nu-glob/src/lib.rs (hunk 8 above)
Comment (sholderbach) [reply crates/nu-glob/src/lib.rs]: Well spotted!

**Review by sholderbach — APPROVED** [review]
I think we should give this one a go! Thanks for the meticulous work @CAD97
